### DISCUSSION TITLE
Proper way to extract AST schema from @babel/types

### DISCUSSION THREAD

**Discussion by jedwards1211** [babel/babel#15276] [Q&A]
I've been extracting the AST schema from `@babel/types` by looking at `NODE_FIELDS` and the `validate` function of each field.  I've [proposed using this in `ast-types`](https://github.com/benjamn/ast-types/issues/823) to ensure that it's using the same AST schema as Babel does, and I do the same thing in https://github.com/codemodsquad/astx.

But my question is, can I rely on this?  Are the properties of `validate` intended to be stable, publically-consumable API or not?  If not, is there any better way, or can we work on making a public API for getting complete information about the AST schema?

**Comment by liuxingbaoyu** [comment]
Seems not, it's an internal implementation detail as I understand it.
What do you want to use it for? Node attribute information? Check node legitimacy? Node parent-child relationship?

**Reply by jedwards1211** [reply]
I want to be able to get the value types of all node fields.  And for fields whose values can be other AST nodes, which AST nodes those are.  `ast-types` allows us to define an arbitrary AST schema and then use its builders/traversal methods on it.  I want to recreate the same schema in `ast-types` as babel uses, so that `recast` (which uses `ast-types` under the hood) behaves properly when operating on an AST parsed by babel.  Currently, `ast-types` declares its own schema for babel, but that's going to lag behind anytime y'all make changes to babel's schema.

The approach I'm currently using works as far as I can tell, I can extract all the information I need.  I just want `@babel/types` to make this metadata available in an official way.

**Reply by jedwards1211** [reply]
Authors of codemod tools (my `astx` or benjamn's `recast`, which is often used with `jscodeshift`) want to be able to support multiple parser/generators so that users can use whichever one they please.  That's why in an ideal world we'd be able to get the AST schema from each parser.

**Reply by liuxingbaoyu** [reply]
`NODE_PARENT_VALIDATIONS` have you tried this, it seems it is exported.

**Reply by jedwards1211** [reply]
@liuxingbaoyu I'm not sure what that's used for but it seems like `ALIAS_KEYS` probably gives me information I can use in the same way?  For example I'm able to use it to declare what node types are considered an `Expression` in the `ast-types` schema

**Reply by liuxingbaoyu** [reply]
It looks like NODE_PARENT_VALIDATIONS is not.
Is the information provided by `ALIAS_KEYS` enough? There is compatibility support for this.

**Reply by liuxingbaoyu** [reply]
Well, I re-looked at the code and the type definitions are being exported, not sure if someone ever decided to expose it, but it is now.
But it would be better if you could use `ALIAS_KEYS`, which is much more stable.

**Reply by jedwards1211** [reply]
Yes `ALIAS_KEYS` works well.  Right now I'm more concerned about properties on `validate` changing.  Right now the only way I've found to determine the type of a node field is:

```ts
    if (validate.type) {
      switch (validate.type) {
        case 'any':
          return {}
        case 'string':
          return builtInTypes.string
        case 'boolean':
          return builtInTypes.boolean
        case 'number':
          return builtInTypes.number
        case 'null':
          return builtInTypes.null
        case 'undefined':
          return builtInTypes.undefined
      }
    }
    if (validate.chainOf) {
      for (const elem of validate.chainOf) {
        const converted = tryConvertValidate(elem)
        if (converted) return converted
      }
    }
    if (validate.each) {
      return [convertValidate(validate.each)]
    }
    if (validate.oneOfNodeTypes) {
      return or(...validate.oneOfNodeTypes.map((type: string) => def(type)))
    }
    if (validate.oneOf) {
      return node?.optional
        ? or(...validate.oneOf, null)
        : or(...validate.oneOf)
    }
    if (validate.shapeOf) {
      return mapValues(validate.shapeOf, (value) =>
        convertValidate(value.validate, value)
      )
    }
    if (validate.oneOfNodeOrValueTypes) {
      return or(
        ...validate.oneOfNodeOrValueTypes.map((type: string) =>
          /^[A-Z]/.test(type) ? def(type) : convertValidate({ type })
        )
      )
    }
```

Where `validate` is coming from

```ts
  for (const [type, fields] of Object.entries(t.NODE_FIELDS)) {
    for (const [field, { validate }] of Object.entries(fields)) {
      ...
    }
  }
```

**Reply by liuxingbaoyu** [reply]
I didn't think of a good way, because even if we add a new verification method here, it may break compatibility.(I don't think it is very likely)
But for now, I think this way is also possible, simple enough. And we can also mention `ast-types` or you if we change these in the future.
